### DISCUSSION TITLE
Brainstorming: Testing distributed NES

### DISCUSSION THREAD

**Discussion by fwc** [nebulastream/nebulastream#134] [Ideas]
# Testing Distributed NES

The following query crashes the worker node.
If the worker is then restarted, the coordinator crashes.

```
Query::from("default_logical")
  .map(Attribute("value") = Attribute("id"))
  .sink(FileSinkDescriptor::create("/tmp/foo.csv", "CSV_FORMAT", "APPEND")
```

While this example could be shrugged of with _"well, we have to improve error handling"_, I think that a distributed setting (i.e. multiple machines talking via network) brings fundamental new failure modes (network, crashes) that need to be _handled_. And the code that does the _handling of failures_ needs to be tested deliberatively.

## Correctness Guarantees

I think what/how we test depends on the correctness guarantees we want to deliver.

I think we should discuss this internally and with leadership to have a strategic goal like e.g. _NES shall have exactly-once guarantees while sustaining network and (some) node failures_.

## Failure Scenarios

The following quotes from [Zeuch CIDR 2020](https://www.nebula.stream/paper/zeuch_cidr20.pdf) describe some failure scenarios, that need to be handled to achieve correctness.

> NebulaStream’s goal is to detect and compensate potentially unreliable nodes in the Fog and Sensor Layer without impacting consistency and availability.

> SLA Centric Processing: NebulaStream’s primary goal is to match user-provided SLAs and QoS constraints with available resources.

Here, some techniques are also mentioned:

> On the Sensor Layer, NES substitutes missing sensor values from broken sensors with nearby sensors, if applicable, or buffers the values during transient connection loss [cite: TinyDB](https://dl.acm.org/doi/abs/10.1145/1061318.1061322). On the Fog Layer, NES extends the [Frontier approach](https://dl.acm.org/doi/abs/10.14778/3231751.3231767), which sends data through multiple network paths to achieve fault-tolerance. Furthermore, data are buffered by upstream operators and replayed in case of an error. On the Cloud Layer, NES extends existing fault-tolerance approaches, e.g., global checkpointing and message broker with fine-grained operator reconfiguration [cite: Ventura @ VLDB PhD Workshop](https://ceur-ws.org/Vol-1882/paper01.pdf).

From the desired correctness guarantees, we should categorise failure scenarios into ones that we intend to handle or do not handle. E.g. we aim to handle _loss of connection between nodes_ and _crashing edge nodes_ but we explicitely do not aim to handle e.g. _disk faults on cloud nodes_.

## Towards Testing for Correctness

I think we can choose an iterative approach here: Over time we get more specific and elaborate, tailoring our approach to the failure scenarios we aim to handle.

I think a good "next thing to do regarding distributed testing" would be an 'elegant' end-to-end test-suite; something like `sqllogictest` + `docker compose`.

Next, this could be combined with some fault injection stuff, e.g. [alexei-led/pumba](https://github.com/alexei-led/pumba) or [chaos-mesh](https://github.com/chaos-mesh/chaos-mesh). It might be nice to combine this with some consistency checker similar to e.g. [elle](https://github.com/jepsen-io/elle), but afaics there is no such thing for streaming.

Afaics, the current state of the art would be _deterministic simulation_ and/or _formal verification_.

![image](https://github.com/user-attachments/assets/22d481a1-90db-4224-ab92-e264c3d1f956)

(x-axis: Investment, y-axis: Returns)

[Peter Alvaro: Orchestrating Chaos Applying Database Research in the Wild](https://youtu.be/YplkQu6a80Q?t=636)

### Deterministic Simulation

The deterministic simulators I'm aware of ([FoundationDB](https://www.youtube.com/watch?v=4fFDFbi3toc), [Stanford's VISORS](https://www.youtube.com/watch?v=prM-0i58XBM) and [RisingWave](https://risingwave.com/blog/deterministic-simulation-a-new-era-of-distributed-system-testing/)) all used coroutine-like code to be able to plug in their simulator-runtime. While the FoundationDB people extended C++ to their [Flow](https://apple.github.io/foundationdb/flow.html) language to get simulation support, the RisingWave people build their simulator ([MadSim](https://github.com/madsim-rs/madsim)) by replacing tokio (async runtime) and tonic (grpc implementation).

One could try to build [_"we have deterministic simulation at home"_](https://knowyourmeme.com/memes/we-have-food-at-home) by re-building the grpc implementation for simulation (like madsim did) in combination with e.g. deterministic multithreading (c.f. [kendo](https://dl.acm.org/doi/10.1145/2528521.1508256), [dthreads](https://dl.acm.org/doi/abs/10.1145/2043556.2043587)), maybe a deterministic allocator (c.f. [KDAlloc](https://dl.acm.org/doi/abs/10.1145/3597926.3604921)), and some refactorings to make NES more deterministic. For this it might be nice to have all inter-node communication via gRPC.

### Formal Verification

As far as I understand, stuff like TLA+ is used to show that some fundamental algorithm like e.g. raft or multi-paxis is correct, which seems to me to be to specific for NES.

The [lightweight formal methods for S3 kv storage node](https://www.amazon.science/publications/using-lightweight-formal-methods-to-validate-a-key-value-storage-node-in-amazon-s3) approach seems a bit more practical where they compared their system to a much smaller "reference" implementation (e.g. simple hashmap instead of kv storage system) by using some [verification tools](https://docs.rs/loom/latest/loom/).

They explicitely mentioned property-based testing, for which [fuzztest](https://github.com/google/fuzztest) seems like a very promising implementation for C++, but I think this is more interesting for single-node testing stuff.

**Comment by fwc** [comment]
# Interesting Material / Link dump

Satarin's [Testing Distributed Systems](https://asatarin.github.io/testing-distributed-systems/) list is very extensive.
Interesting Sections: [Deterministic Simulation](https://asatarin.github.io/testing-distributed-systems/#deterministic-simulation), [Chaos Engineering](https://asatarin.github.io/testing-distributed-systems/#chaos-engineering), [Fuzzing](https://asatarin.github.io/testing-distributed-systems/#fuzzing), [Performance and Benchmarking](https://asatarin.github.io/testing-distributed-systems/#performance-and-benchmarking), [Rising Wave](https://asatarin.github.io/testing-distributed-systems/#risingwave), [Wallaroo Labs](https://asatarin.github.io/testing-distributed-systems/#wallaroo-labs)

Also maybe interesting:

- https://side-eye.io/
- Meiklejohn: Partisan: Enabling Real-World Protocol Evaluation

**Comment by fwc** [comment]
# What would be the consistency Model / Test invariant(s) / ...?

For testing (distributed) Databases, there are theoretical [consistency models](https://jepsen.io/consistency). Software like [jepsen](https://jepsen.io/) injects faults and later detects violations of the guarantees required by the models.

Put differently: During e.g. black-box testing, we check if externally observable invariants are violated (e.g. a segfault). Which invariants would be interesting for NES?

**Reply by fwc** [reply]
Delivery guarantee / Exactly Once & friends.

**Comment by fwc** [comment]
# Rising Wave / MadSim

Rising Wave advertises itself as an _SQL Database for Streaming Data._ and is written in Rust. They use deterministic simulation for testing, for which they built [MadSim](https://github.com/madsim-rs/madsim).

Afaics, MadSim works by replacing [tokio](https://github.com/tokio-rs/tokio) (async runtime) and [tonic](https://github.com/hyperium/tonic) (gRPC implementation), so that scheduling and communication are in full control of the simulator.

**Comment by keyseven123** [comment]
I am not sure how others feel, but for me it is quite difficult to understand what you think we should do. Maybe it would be cool to have a list of pros and cons for each idea/way of testing?
